flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java

```diff
@@ -0,0 +1,312 @@
+            // So we explicitly store this and when we do a persist call we make use of it.
+            // This we have raised a bug in ABFS hadoop driver side. Once fixed this will not be
+            // needed. So it should be ok to put this in side the 'if' check.
+            initialFileSize = fs.getFileStatus(tempFile).getLen();
+        }
+        LOG.debug("Created a new OS for appending {}", tempFile);
+    }
+
+    private void truncate(FileSystem fs, HadoopFsRecoverable recoverable) throws IOException {
+        Path renameTempPath = new Path(tempFile.toString() + ".rename");
+        try {
+            LOG.info(
+                    "Creating the temp rename file {} for truncating the tempFile {}",
+                    renameTempPath,
+                    tempFile);
+            FSDataOutputStream fsDataOutputStream = fs.create(renameTempPath);
+            LOG.info("Opening the tempFile {} for truncate", tempFile);
+            FSDataInputStream fsDis = fs.open(tempFile);
+            // 2 MB buffers. TODO : Make this configurable
+            long remaining = recoverable.offset();
+            byte[] buf = null;
+            long dataWritten = 0;
+            int readBytes = -1;
+            while (remaining != 0) {
+                if (minBufferLength < remaining) {
+                    buf = new byte[minBufferLength];
+                } else {
+                    buf = new byte[(int) remaining];
+                }
+                readBytes = fsDis.read(buf, 0, buf.length);
+                if (readBytes != -1) {
+                    remaining -= readBytes;
+                    LOG.info("Bytes remaining to read {}", remaining);
+                    fsDataOutputStream.write(buf, 0, readBytes);
+                    dataWritten += readBytes;
+                    LOG.info("Successfully wrote {} bytes of data", dataWritten);
+                } else {
+                    LOG.debug("Reached the end of the file");
+                    remaining = 0;
+                }
+            }
+            // TODO : Support intermediate flush?
+            LOG.info("Closing the temp rename file {}", renameTempPath);
+            fsDataOutputStream.close();
+            if (fsDis != null) {
+                LOG.debug("Closing the input stream");
+                fsDis.close();
+            }
+        } catch (IOException e) {
+            LOG.error(
+                    "Unable to recover. Exception while trying to truncate the temp file {}",
+                    tempFile);
+            // We cannot recover. This we can control if user does not want this??
+            throw e;
+        }
+        try {
+            LOG.info("Deleting the actual temp file {}", tempFile);
+            fs.delete(tempFile, false);
+        } catch (IOException e) {
+            LOG.error("Unable to recover. Error while deleting the temp file {}", tempFile);
+            // unable to recover.
+            throw e;
+        }
+        rename(fs, renameTempPath);
```

Comment:
What if there is a fail in between these steps of delete and rename?  Later how it will be possible to rollback.  The truncate should be able to handle such case?

Reply:
In HDFS case we try to trucnate and if that is still failing we were throwing exceptions. So I think the same is happening here. This recovery will be a failure. Also as I said truncate happens only in case where we move to an older recovery checkpoint.

Reply:
In our testing with DefaultRollingpolicy - the recovery is not really possible as the inprogress that was marked would have already been committed in the next checkpoint. Only with policy as Checkpoint based rolling policy we can achieve this . So basically when moving to an older checkpoint - I believe  yes if we fail in a delete and rename step it is a non recoverable error but we can always move to latest checkpoint. But am not very sure how this recovery will be used in case of FS based sinks.

Reply:
Now I have handled this case by checking if the '.rename' file is available and if that is really of the expected length just go with rename. This should solve the problem of where the failure happens between delete and rename.